bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/SpanImpl.kt

```diff
@@ -317,4 +346,180 @@ public class SpanImpl internal constructor(
             }
         }
     }
+
+    /**
+     * Represents a possible "block" to a Span being ended naturally. Any conditions that are active
+     * when [SpanImpl.end] is called will prevent the span from being batched and processed, until
+     * all of the [Condition]s are either cancelled, timed-out, or closed. A condition is considered
+     * "inactive" until it is [upgrade]ed, at which point the [Condition.close] will cause the span
+     * end-time to be adjusted.
+     *
+     * These are a way to deal with cases where a span "might have child spans at some point in the
+     * future, where the parent span end time should match the last of these children".
+     */
+    public interface Condition {
+        /**
+         * Close this [Condition] possibly releasing the underlying Span for batching and
+         * processing. The Span end time may be adjusted to [endTime] if it is strictly after the
+         * existing Span end time.
+         */
+        public fun close(endTime: Long = SystemClock.elapsedRealtimeNanos())
+
+        /**
+         * Attempt to upgrade this [Condition]. A Condition may only be upgraded only be upgraded
+         * once, and must still be valid (not closed, or cancelled) in order to be upgraded.
+         *
+         * @return the [SpanContext] that the [Condition] was upgraded to, or `null` if this
+         * [Condition] cannot be upgraded
+         */
+        public fun upgrade(): SpanContext?
+
+        /**
+         * Cancel this [Condition] possibly releasing the underlying Span for batching and
+         * processing. This function has the same effect as the [Condition] passing its timeout.
+         * The [Span] this condition is blocking will not be altered by this function.
+         */
+        public fun cancel()
+    }
+
+    private class ConditionImpl(
+        private val span: SpanImpl,
+        override val target: Long,
+    ) : Condition, Timeout {
+        private var isValid = true
+        private var isUpgraded = false
+
+        override fun close(endTime: Long) {
+            releaseCondition {
+                if (isUpgraded) {
+                    span.state.endTime = max(endTime, span.state.endTime)
+                }
+            }
+        }
```

Comment:
What happens if a span is discarded after being blocked? It seems like it could still potentially be sent for processing?

Reply:
That's a really good point, discards should be cancelling all of the open `Condition`s. I'll amend 👍